src/tools/rbd/action/Export.cc

```diff
@@ -86,12 +365,156 @@ class C_Export : public Context
   SimpleThrottle &m_throttle;
   librbd::Image &m_image;
   bufferlist m_bufferlist;
+  uint64_t m_dest_offset;
   uint64_t m_offset;
   uint64_t m_length;
   int m_fd;
 };
 
-static int do_export(librbd::Image& image, const char *path, bool no_progress)
+static int do_export_v2(librbd::Image& image, librbd::image_info_t &info, int fd,
+		        uint64_t period, int max_concurrent_ops, utils::ProgressContext &pc)
+{
+  int r = 0;
+  // header
+  bufferlist bl;
+  bl.append(utils::RBD_IMAGE_BANNER_V2);
+
+  // size in header
+  uint64_t size = info.size;
+  ::encode(size, bl);
```

Comment:
Perhaps this should be omitted and instead encoded in the "snap" section. If I have an image w/ snapshots, the first snapshot could be size X, the second could be size Y, and the HEAD revision could be size Z. When processing the first snapshot, you would need to resize the image down unnecessarily. Therefore, instead, just create a zero-sized image until you get to the first diff section which will tell you the current image size.